primitives/arithmetic/src/fixed128.rs

```diff
@@ -0,0 +1,675 @@
+			rhs = rhs.saturating_mul(-1);
+		}
+
+		U256::from(lhs)
+			.checked_mul(U256::from(rhs))
+			.and_then(|n| n.checked_div(U256::from(DIV)))
+			.and_then(|n| TryInto::<i128>::try_into(n).ok())
+			.map(|n| Self(n * signum))
+	}
+
+	/// Checked div. Same semantic to `num_traits::CheckedDiv`.
+	pub fn checked_div(&self, rhs: &Self) -> Option<Self> {
+		if rhs.0.signum() == 0 {
+			return None;
+		}
+		if self.0 == 0 {
+			return Some(*self);
+		}
+
+		let signum = self.0.signum() / rhs.0.signum();
+		let mut lhs: i128 = self.0;
+		if lhs.is_negative() {
+			lhs = lhs.saturating_mul(-1);
+		}
+		let mut rhs: i128 = rhs.0.saturated_into();
+		if rhs.is_negative() {
+			rhs = rhs.saturating_mul(-1);
+		}
+
+		U256::from(lhs)
+			.checked_mul(U256::from(DIV))
+			.and_then(|n| n.checked_div(U256::from(rhs)))
+			.and_then(|n| TryInto::<i128>::try_into(n).ok())
+			.map(|n| Self(n * signum))
+	}
+
+	/// Checked mul for int type `N`.
+	pub fn checked_mul_int<N>(&self, other: &N) -> Option<N>
+	where
+		N: Copy + TryFrom<i128> + TryInto<i128>,
+	{
+		N::try_into(*other).ok().and_then(|rhs| {
+			let mut lhs = self.0;
+			if lhs.is_negative() {
+				lhs = lhs.saturating_mul(-1);
+			}
+			let mut rhs: i128 = rhs.saturated_into();
+			let signum = self.0.signum() * rhs.signum();
+			if rhs.is_negative() {
+				rhs = rhs.saturating_mul(-1);
+			}
+
+			U256::from(lhs)
+				.checked_mul(U256::from(rhs))
+				.and_then(|n| n.checked_div(U256::from(DIV)))
+				.and_then(|n| TryInto::<i128>::try_into(n).ok())
+				.and_then(|n| TryInto::<N>::try_into(n * signum).ok())
+		})
+	}
+
+	/// Checked mul for int type `N`.
+	pub fn saturating_mul_int<N>(&self, other: &N) -> N
+	where
+		N: Copy + TryFrom<i128> + TryInto<i128> + Bounded,
+	{
+		self.checked_mul_int(other).unwrap_or_else(|| {
+			N::try_into(*other)
+				.map(|n| n.signum())
+				.map(|n| n * self.0.signum())
+				.map(|signum| {
+					if signum.is_negative() {
+						Bounded::min_value()
+					} else {
+						Bounded::max_value()
+					}
+				})
+				.unwrap_or(Bounded::max_value())
+		})
+	}
+
+	/// Checked div for int type `N`.
+	pub fn checked_div_int<N>(&self, other: &N) -> Option<N>
+	where
+		N: Copy + TryFrom<i128> + TryInto<i128>,
+	{
+		N::try_into(*other)
+			.ok()
+			.and_then(|n| self.0.checked_div(n))
+			.and_then(|n| n.checked_div(DIV))
+			.and_then(|n| TryInto::<N>::try_into(n).ok())
+	}
+
+	pub fn zero() -> Self {
+		Self(0)
+	}
+
+	pub fn is_zero(&self) -> bool {
+		self.0 == 0
+	}
+
+	/// Saturating absolute value. Returning MAX if `parts` == i128::MIN instead of overflowing.
+	pub fn saturating_abs(&self) -> Self {
+		if self.0 == i128::min_value() {
+			return Fixed128::max_value();
+		}
+
+		if self.0.is_negative() {
+			Fixed128::from_parts(self.0 * -1)
+		} else {
+			*self
+		}
+	}
+
+	pub fn is_positive(&self) -> bool {
+		self.0.is_positive()
+	}
+
+	pub fn is_negative(&self) -> bool {
+		self.0.is_negative()
+	}
+}
+
+impl Saturating for Fixed128 {
```

Comment:
```suggestion
/// Note that this is a standard, _potentially-panicking_, implementation. Use `Saturating` trait
/// for safe addition.
impl ops::Add for Fixed128 {
	type Output = Self;

	fn add(self, rhs: Self) -> Self::Output {
		Self(self.0 + rhs.0)
	}
}

/// Note that this is a standard, _potentially-panicking_, implementation. Use `Saturating` trait
/// for safe subtraction.
impl ops::Sub for Fixed128 {
	type Output = Self;

	fn sub(self, rhs: Self) -> Self::Output {
		Self(self.0 - rhs.0)
	}
}

impl Saturating for Fixed128 {
```

Any reason not to include basic Add/Sub traits? Makes it usable in many scenarios, especially tests.

Reply:
we just don't want to introduce any new possibilities of panics.

for tests we are currently use `value.checked_add(other).unwrap()`

Reply:
Practically speaking though, not implementing these traits can be a major pain in the ass for regular usage. My use case specifically was using `assert_eq_error_rate` which needed the type to support `Add`. I mean there are other ways to fix this problem, but I dont think you are really causing any more or less problems implementing these very standard traits.

Reply:
Maybe add it with the `test` cfg